pkg/cfn/manager/tasks.go

```diff
@@ -1,158 +1,233 @@
-// will be returned as a slice as soon as one of the tasks or group
-// of tasks is completed
-func (c *StackCollection) CreateClusterWithNodeGroups(onlySubset sets.String) []error {
-	if errs := c.RunSingleTask(Task{c.CreateCluster, nil}); len(errs) > 0 {
-		return errs
+// Do will run through the set in the backround, it may return an error immediately,
+// or eventually write to the errs channel; it will close the channel once all tasks
+// are completed
+func (t *TaskTree) Do(allErrs chan error) error {
+	if t.Len() == 0 || t.DryRun {
+		logger.Debug("no actual tasks")
+		close(allErrs)
+		return nil
 	}
 
-	return c.CreateAllNodeGroups(onlySubset)
-}
+	errs := make(chan error)
 
-// CreateAllNodeGroups runs all tasks required to create the node groups;
-// any errors will be returned as a slice as soon as one of the tasks
-// or group of tasks is completed
-func (c *StackCollection) CreateAllNodeGroups(onlySubset sets.String) []error {
-	errs := []error{}
-	appendErr := func(err error) {
-		errs = append(errs, err)
+	if t.Parallel {
+		go doParallelTasks(errs, t.tasks)
+	} else {
+		go doSequentialTasks(errs, t.tasks)
 	}
 
-	createAllNodeGroups := []Task{}
-	for i := range c.spec.NodeGroups {
-		ng := c.spec.NodeGroups[i]
-		if onlySubset != nil && !onlySubset.Has(ng.Name) {
-			continue
+	go func() {
+		defer close(allErrs)
+		for err := range errs {
+			allErrs <- err
 		}
-		t := Task{
-			Call: c.CreateNodeGroup,
-			Data: ng,
-		}
-		createAllNodeGroups = append(createAllNodeGroups, t)
-	}
-	if Run(appendErr, createAllNodeGroups...); len(errs) > 0 {
-		return errs
-	}
+		// for {
+		// 	select {
+		// 	case err, open := <-errs:
+		// 		if !open {
+		// 			return
+		// 		}
+		// 		allErrs <- err
+		// 	}
+		// }
+	}()
 
 	return nil
 }
 
-// CreateOneNodeGroup runs a task to create a single node groups;
-// any errors will be returned as a slice as soon as the tasks is
-// completed
-func (c *StackCollection) CreateOneNodeGroup(ng *api.NodeGroup) []error {
-	return c.RunSingleTask(Task{
-		Call: c.CreateNodeGroup,
-		Data: ng,
-	})
-}
+// DoAllSync will run through the set in the foregounds and return all the errors
+// in a slice
+func (t *TaskTree) DoAllSync() []error {
+	if t.Len() == 0 || t.DryRun {
+		logger.Debug("no actual tasks")
+		return nil
+	}
+
+	errs := make(chan error)
 
-// DeleteAllNodeGroups deletes all nodegroups without waiting
-func (c *StackCollection) DeleteAllNodeGroups(call taskFunc) []error {
-	nodeGroupStacks, err := c.DescribeNodeGroupStacks()
-	if err != nil {
-		return []error{err}
+	if t.Parallel {
+		go doParallelTasks(errs, t.tasks)
+	} else {
+		go doSequentialTasks(errs, t.tasks)
 	}
 
-	errs := []error{}
-	for _, s := range nodeGroupStacks {
-		if err := c.DeleteNodeGroup(getNodeGroupName(s)); err != nil {
-			errs = append(errs, err)
-		}
+	allErrs := []error{}
+	for err := range errs {
+		allErrs = append(allErrs, err)
 	}
+	return allErrs
+	// for {
```

Comment:
@rndstr would you happen to know if `for x := range <channel>` is exactly the same as above? It seems like it is, but I'm not 100% sure.

Reply:
yep looks the same to me.
(with the caveat that if you were to have code after that block you'd need a `break <label>` to get out of the for _and_ select  I think)

Reply:
Yeah... I think select just let's you handle multiple channels nicely, it makes sense to have `for/range` shorthand syntax for a single channel.